src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,215 @@
+                            using (var fileStream = new FileStream(item.targetPath, FileMode.Create))
+                            {
+                                item.stream.CopyTo(fileStream);
+                            }
+                        }
+                        catch (IOException exception) when (retryCount < 3)
+                        {
+                            tracer.Warn($"Exception caught: {exception.Message}, on retry count {retryCount}, Retrying");
+                            retryCount++;
+                        }
+                    }               
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 8,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount * 8,
+                    CancellationToken = cancellationToken,
+                });
+
+            batchItemsBlock.LinkTo(fetchStream, new DataflowLinkOptions() { PropagateCompletion = true });
+            fetchStream.LinkTo(downloadBlock, new DataflowLinkOptions() { PropagateCompletion = true });
+
+            await batchItemsBlock.SendAllAndCompleteAsync(fileItems, downloadBlock, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<Stream> DownloadFileFromContainerAsync(
+            (long, string) containerIdAndRoot,
+            Guid scopeIdentifier,
+            FileContainerHttpClient containerClient,
+            FileContainerItem item,
+            CancellationToken cancellationToken)
+        {
+            Stream responseStream = await AsyncHttpRetryHelper.InvokeAsync(
+                async () =>
+                {
+                    Stream internalResponseStream = await containerClient.DownloadFileAsync(containerIdAndRoot.Item1, item.Path, cancellationToken, scopeIdentifier).ConfigureAwait(false);
+                    return internalResponseStream;
+                },
+                maxRetries: 5,
+                cancellationToken: cancellationToken,
+                tracer: this.tracer,
+                continueOnCapturedContext: false
+                );
+
+            return responseStream;
+        }
+
+        private string ResolveTargetPath(string rootPath, FileContainerItem item, string artifactName)
+        {
+            if(item.Path.Length > artifactName.Length)
+            {
+                artifactName += "/";
+            }
+            var itemPathWithoutDirectoryPrefix = item.Path.Replace(artifactName, "");
+            var absolutePath = Path.Combine(rootPath, itemPathWithoutDirectoryPrefix);
+            return absolutePath;
+        }
+
+        private List<FileContainerItem> GetFilteredItems(List<FileContainerItem> items, IEnumerable<Func<string, bool>> minimatchFuncs, string artifactName)
+        {
+            List<FileContainerItem> filteredItems = new List<FileContainerItem>();
+            int index = artifactName.Length;
```

Comment:
One more: It is possible to have an artifact where artifactName is different that the container root path. This I think isn't possible when we directly use the publish build artifact task but is possible when the artifact is associated using agent command. I don't have data as to how many such type of artifacts are there, but that is something that you might need to look out for.

Reply:
Thanks for the information. If that's the case, we should change the ParseContainerId function as well??

Reply:
Depends on how one wants to handle it. In RM we download the artifacts into folder name matching the container folder name instead of artifact folder name. So ParseContainerId works correctly in our case.

Reply:
Perfect. Will use that logic here as well. Thanks.